Make 'kw maintainers' work from any directory

MatheusBernardino opened the pull request:
This PR makes  'kw maintainers' work  from any directory. It fixes #28 

It does that looking for he kernel tree root, cd-ing to it before running get_maintainers and cd-ing back after.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -47,6 +47,31 @@ function print_files_authors()
     done
 }
 
+function is_linux_tree_root()
```
how about put this function on "commons.sh"? This can be useful for a new version of the function handler for checkpatch (see #40)

MatheusBernardino replied:
Seems good! But don't you think it would be nicer to leave the function here, for now, and move it in a PR for #40 ? (Just thinking about that idea you told us of "Do what is sufficient for now, not for the future", you know?)

MatheusBernardino commented:
Rebased and added tests, as requested. Also, improved get_maintainers.pl behavior to handle even more corner cases, as described in the commit message.

MatheusBernardino commented:
> First of all, thanks for your PR! I really enjoyed reviewing this one :-)
> 
> I just have two comments:
> 
>     1. Rebase you code with the branch "unstable"
> 
>     2. Could you add some tests?

Rebased and added tests, as requested. Also, improved get_maintainers.pl behavior to handle even more corner cases, as described in the commit message.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -47,40 +48,168 @@ function print_files_authors()
     done
 }
 
+# Returns true if given dir is a kernel tree root. Argument is the dir.
```
Use spaces instead of tabs.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -47,40 +48,168 @@ function print_files_authors()
+  if [[ -f "${DIR}/COPYING" &&
+  		-f "${DIR}/CREDITS" &&
+		-f "${DIR}/Kbuild" &&
+		-e "${DIR}/MAINTAINERS" &&
```
Why you use '-e' instead of '-f' here?

MatheusBernardino replied:
Oh, nice question. The get_maintainer.pl script uses this exactly same tests to check for a kernel root. So I used '-e' instead of '-f' to keep this function consistent with get_maintainer.pl

MatheusBernardino replied:
Oh, nice question. The get_maintainer.pl script uses this exactly same tests to check for a kernel root. So I used '-e' instead of '-f' to keep this function consistent with get_maintainer.pl.

MatheusBernardino replied:
Oh, nice question. The get_maintainer.pl script uses this exactly same tests to check for a kernel root. So I used '-e' instead of '-f' to keep this function consistent with get_maintainer.pl.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -47,40 +48,168 @@ function print_files_authors()
+
+	# If file is not a patch and outside a kernel tree, it must be an user's
+	# mistake. Although get_maintainer.pl can handle this, and would print
+	# the kernel root maintainer, linux torvalds, it's better to abort
```
I did not understood what you meant by "linux torvalds", but in any case, I prefer no name references in the code.

rodrigosiqueira commented on tests/get_maintainer_wrapper_test.sh:
```diff
@@ -3,60 +3,145 @@
+John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org"
+
+
+
```
get rid of the extras spaces. Just one line is enough.

rodrigosiqueira commented:
Hi, first of all: thanks a lot for your PR :)

We are close to finish this. I have some comments:

1. Try command below produce an infinity loop:

```
kw m ../la/lu.txt
```

2. Take care with tabs, and white spaces. Take a look in your editor (I always recommend use vim ;). We use space as default in kworkflow.

3. Keep the 'if' condition patterns. Sometimes you use '[]', '[[ ]]' and sometimes you use nothing. I recommend the use of '[[ ]]' every time. (Actually... I'm surprised that any error message was not pop up)

4. I really liked these functions: is_kernel_root(), find_kernel_root(), and is_a_patch(). It could be better to move all of them to `commons.sh` as the first commit in this patchset. Then, you replace the include of `miscellaneous.sh` by `commons.sh`. These functions could be used to improve the `checkpatch` in the future, because of this I recommend you to replace this in the `commons`.

5. There is some weird behavior associated with the new tests. I executed the test two or three times, and I got the following error:

```
Running test [get_maintainer_wrapper_test]
=========================================================
testPrintFileAuthorForFile
testPrintFileAuthorForDir
testLinuxRootCheck
testGetMaintainers
./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
ASSERT:Expecting return:\n> =========================================================
> HERE:
> Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> HERE:\n(Calling lkr from outside lkr)
./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
ASSERT:Expecting return:\n> =========================================================
> HERE:
> Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> HERE:\n(Calling lkr from lkr)
./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/fs/' doesn't appear to be a patch.  Add -f to options?
ASSERT:Expecting return:\n> =========================================================
> HERE:
> John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org\nBut got:\n> =========================================================
> HERE:\n(Calling lkr/fs from lkr)
./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
ASSERT:Expecting return:\n> =========================================================
> HERE:
> Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> HERE:\n(Calling lkr from lkr/fs)
./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/fs/' doesn't appear to be a patch.  Add -f to options?
ASSERT:Expecting return:\n> =========================================================
> HERE:
> John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org\nBut got:\n> =========================================================
> HERE:\n(Calling lkr/fs from lkr/fs)

Ran 4 tests.

FAILED (failures=5)
```
However... for some reason, it started to pass.

MatheusBernardino commented:
Hi, @rodrigosiqueira . I finally got back to this PR (really sorry for such a long delay). I corrected most of the things but there are some I want a new feedback from you, since it has been such a long time since the review.

>     1. Try command below produce an infinity loop:
> ```
> kw m ../la/lu.txt
> ```

I couldn't reproduce it here. Can you make me a more concrete example?

>     2. Keep the 'if' condition patterns. Sometimes you use '[]', '[[ ]]' and sometimes you use nothing. I recommend the use of '[[ ]]' every time. (Actually... I'm surprised that any error message was not pop up)

Ok, I can replace the '[]' by '[[]]' (since there seems to be a compatibility), but as for the `if ! $bla;`, I would have to use something like `if [[ $bla -ne true ]];`, which seems to complicate the statement. Do you think it's a good idea? Or do you know if there is any other way to do a "if not" using '[[]]'?

Also, I'm reading a lot of people recommending to use '[]' instead of '[[]]', since the last is bash-only and the first is POSIX (although '[[]]' seems to be a great enhancement over '[]'). Would we like to make kw more portable in the near future, so that it can be installed on zsh and others?

>     3. I really liked these functions: is_kernel_root(), find_kernel_root(), and is_a_patch(). It could be better to move all of them to `commons.sh` as the first commit in this patchset. Then, you replace the include of `miscellaneous.sh` by `commons.sh`. These functions could be used to improve the `checkpatch` in the future, because of this I recommend you to replace this in the `commons`.
> 

Since the review, we changed a lot of things related to `commons` and `miscellaneous` (the latter doesn't even exist anymore), but the overall idea is still doable. Even so, `commons` became a file containing exclusively local/global configurations handlers. And, IMO, it would be a very good design to leave it that way, i.e., not to mix other things in this file (I would even suggest to rename `commons` to `config_handler`, or something like this). I know `find_kernel_root` may be used in another PR to fix issue #40, but IMO, we should leave it at `get_maintainger_wrapper` for now, so we can get this patch merged, and when someone (possibly me :) goes to fix #40, we discuss were to move this function to. What do you think?

>     4. There is some weird behavior associated with the new tests. I executed the test two or three times, and I got the following error:
> 
> 
> ```
> Running test [get_maintainer_wrapper_test]
> =========================================================
> testPrintFileAuthorForFile
> testPrintFileAuthorForDir
> testLinuxRootCheck
> testGetMaintainers
> ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
> ASSERT:Expecting return:\n> =========================================================
> > HERE:
> > Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> > HERE:\n(Calling lkr from outside lkr)
> ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
> ASSERT:Expecting return:\n> =========================================================
> > HERE:
> > Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> > HERE:\n(Calling lkr from lkr)
> ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/fs/' doesn't appear to be a patch.  Add -f to options?
> ASSERT:Expecting return:\n> =========================================================
> > HERE:
> > John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org\nBut got:\n> =========================================================
> > HERE:\n(Calling lkr/fs from lkr)
> ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
> ASSERT:Expecting return:\n> =========================================================
> > HERE:
> > Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> > HERE:\n(Calling lkr from lkr/fs)
> ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/fs/' doesn't appear to be a patch.  Add -f to options?
> ASSERT:Expecting return:\n> =========================================================
> > HERE:
> > John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org\nBut got:\n> =========================================================
> > HERE:\n(Calling lkr/fs from lkr/fs)
> 
> Ran 4 tests.
> 
> FAILED (failures=5)
> ```
> 
> However... for some reason, it started to pass.

Again, I couln't reproduce it here :( Was it just the first time you executed the test? Do you have any clue of what happened or how I may be able to replicate it? (By the output, it may be something related to the '[]' / '[[]]' / 'nothing' thing, we talked earlier.)

MatheusBernardino commented:
> I just did:
> 
> ```
> cd /tmp
> kw m ../la/lu.txt
> ```
> 
> I tried it again, and I still get this look.
> 

Oh, now I was able to replicate! I was creating the `la` and `lu` dirs and `lu.txt` file, but the infinity loop happens exactly when the dirs/files does not exits. Thanks, I'll fix that!

> For me using [[ -eq ]] is totally fine. It is just a syntax pattern. Anyway, you can try other alternatives. 

Ok. So can I use `if [[ $BLA -eq true]]` to test wether a variable is true? It stills seems a little too verbose to me, since we could just use `if $BLA` (But I really don't know about the disadvantages of using it. Maybe the way you suggested, with '[[]]' is more secure?).

> How about using one of these patterns:
> 
> ```
> command && another_command
> command || other_command
> ```

Yes, I'm already using this patterns, as in `if ! $is_file_a_patch && ! $is_file_inside_kernel_tree; then`, for example. But I understood you prefer not to have if's without '[]' or '[[]]', is that right? Or should I leave if's like the above as they are?  


> Yes, I know. However, take a look at the top of our files, and you’ll see “#!/bin/bash” (i.e., we already rely on bash). Additionally, it is a real pain to make a big set os shell scripts portable, and IMHO we should avoid it. Finally, the shebang forces the current terminal to use bash.
> 

I know we use `#!/bin/bash`, bash style functions and other things that would have to me changed for a POSIX kw, but just wanted to know wether that's a plan for the future or not. I agree with you that it would be too much hard work and it's not worthy for now.

> > > ```
> > > 4. There is some weird behavior associated with the new tests. I executed the test two or three times, and I got the following error:
> > > ```
> > > ```
> > > Running test [get_maintainer_wrapper_test]
> > > =========================================================
> > > testPrintFileAuthorForFile
> > > testPrintFileAuthorForDir
> > > testLinuxRootCheck
> > > testGetMaintainers
> > > ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> > > scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
> > > ASSERT:Expecting return:\n> =========================================================
> > > > HERE:
> > > > Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> > > > HERE:\n(Calling lkr from outside lkr)
> > > ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> > > scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
> > > ASSERT:Expecting return:\n> =========================================================
> > > > HERE:
> > > > Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> > > > HERE:\n(Calling lkr from lkr)
> > > ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> > > scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/fs/' doesn't appear to be a patch.  Add -f to options?
> > > ASSERT:Expecting return:\n> =========================================================
> > > > HERE:
> > > > John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org\nBut got:\n> =========================================================
> > > > HERE:\n(Calling lkr/fs from lkr)
> > > ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> > > scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/' doesn't appear to be a patch.  Add -f to options?
> > > ASSERT:Expecting return:\n> =========================================================
> > > > HERE:
> > > > Jane Doe <jane@email.com>,kernel@list.org\nBut got:\n> =========================================================
> > > > HERE:\n(Calling lkr from lkr/fs)
> > > ./src/get_maintainer_wrapper.sh: line 178: [: is_a_patch: unary operator expected
> > > scripts/get_maintainer.pl: file '/home/siqueira/Documents/Repositories/kworkflow/tests/.tmp/fs/' doesn't appear to be a patch.  Add -f to options?
> > > ASSERT:Expecting return:\n> =========================================================
> > > > HERE:
> > > > John Doe <john@email.com>,Jane Doe <jane@email.com>,fs@list.org,kernel@list.org\nBut got:\n> =========================================================
> > > > HERE:\n(Calling lkr/fs from lkr/fs)
> > > 
> > > Ran 4 tests.
> > > 
> > > FAILED (failures=5)
> > > ```
> > > However... for some reason, it started to pass.
> > 
> > 
> > Again, I couln't reproduce it here :( Was it just the first time you executed the test? Do you have any clue of what happened or how I may be able to replicate it? (By the output, it may be something related to the '[]' / '[[]]' / 'nothing' thing, we talked earlier.)
> 
> Sorry, I don't remember about this as well. I recommend you to rebase your branch with the unstable branch, and send a new version. I will test it again ;)

The branch in this PR is already rebased with unstable :) If you can test and give me some feedback on this test error, I would really appreciate :)

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+      fi
+    done
+  elif [[ -f $FILE_OR_DIR ]]; then
+    files+=($FILE_OR_DIR)
```
How about adding an else operation for the case that something weird is passed in $FILE_OR_DIR? Another question, how about check if the user has permission to the directory and file?
IMHO this can be good to provide precise information for users.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
-
+# Prints the authors of a given file or files inside a given dir.
+#
+# @parameter The argument is a file or directory path
```
How about use "@FILE_OR_DIR The ..." instead of @parameter? The Kernel documentation use this style.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
-        done
-    elif [[ -f $FILE_OR_DIR ]]; then
-        files+=($FILE_OR_DIR)
+  # TODO: currently only authors found in a single line MODULE_AUTHOR
```
How about create an issue for this?

MatheusBernardino replied:
Ok, created and comment removed from code.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+
+# Checks if a directory is a kernel tree root
+#
+# @parameter A directory path
```
How about "@DIR"?

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+        -d "${DIR}/ipc" &&
+        -d "${DIR}/kernel" &&
+        -d "${DIR}/lib" &&
+        -d "${DIR}/scripts" ]]; then
```
I’m a little bit worried about this part since it makes the “kw m” fragile. If the Linux Kernel changes a single file, “kw m” will fail. Also, this hardcoded option could be a little bit more flexible.

My suggestions:
1. Add three or four files for checking if the directory is a kernel root or not. I suggest you inspect for get_maintainers.pl, MAINTAINERS, codestyle.pl, and kernel directory.
2. Add a second if, that loads an extra variable from etc/kworkflow.config. This variable could have other files specified by the user; with this, if ‘kw m’ fails to match the root kernel, the user can add a fine-grained level for an extra check.

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
-        fi
+# Finds the root of the linux kernel repo containing the given file
+#
+# @parameter The argument is a directory of file path
```
@FILE_OR_DIR

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+  fi
+
+  # Find the kernel tree root
+  if is_kernel_root "$current_dir"; then
```
```if [[ $(is_kernel_root "$current_dir") ]]; then``` ?

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+  else
+    while [[ "$current_dir" != "/" ]]; do
+      current_dir="$(dirname $current_dir)"
+      if is_kernel_root "$current_dir"; then
```
Use [[ ]]?

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+        break
+      fi
     done
+  fi
```
newline

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+# True if given path is a patch file and false otherwise.
+function is_a_patch
+{
+  local -r FILE_PATH="$@"
```
newline

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
 
+# Checks if the given path is a patch file
+#
+# @parameter The argument is the path
```
@FILE_PATH

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+  )
+
+  for expected_str in "${PATCH_EXPECTED_STRINGS[@]}"; do
+    if ! grep -F "$expected_str" "$FILE_PATH" &> /dev/null; then
```
[[ ]]

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+#
+# @parameter First argument can be "--authors" or "a" to enable authors
+#            printing. It is optional.
+# @parameter Second argument is the given file/dir. It is optional and, if not
```
Just @FILE_OR_DIR and merge the comments?

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
-  if [ $? -ne 0 ]; then
-    return $?
+  # if given path is not a patchfile, add -f to get_maintainer.pl options
+  if ! is_a_patch "$FILE_OR_DIR"; then
```
[[ ]]

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+    is_file_inside_kernel_tree=false
+    # fallback: try to find kernel root at working path
+    kernel_root="$(find_kernel_root $original_working_dir)"
+  fi
```
newline

rodrigosiqueira commented on src/get_maintainer_wrapper.sh:
```diff
@@ -1,90 +1,240 @@
+  # If file is not a patch and outside a kernel tree, it must be an user's
+  # mistake. Although get_maintainer.pl can handle this, it's better to abort
+  # because it is most likely a user's mistake. So better let the user know.
+  if ! $is_file_a_patch && ! $is_file_inside_kernel_tree; then
```
[[ ]]

rodrigosiqueira commented on tests/get_maintainer_wrapper_test.sh:
```diff
@@ -3,60 +3,141 @@
+  if [[ "$ret" != "$CORRECT_TMP_MSG" ]]; then
+    got_prefixed=$(prefix_multiline "$ret")
+    fail "Expecting return:\n$expected_tmp_prefixed\nBut got:\n$got_prefixed\n(Calling lkr from outside lkr)"
+  fi
```
newline

rodrigosiqueira commented on tests/get_maintainer_wrapper_test.sh:
```diff
@@ -3,60 +3,141 @@
+  if [[ "$ret" != "$CORRECT_TMP_MSG" ]]; then
+    got_prefixed=$(prefix_multiline "$ret")
+    fail "Expecting return:\n$expected_tmp_prefixed\nBut got:\n$got_prefixed\n(Calling lkr from lkr)"
+  fi
```
newline

rodrigosiqueira commented on tests/get_maintainer_wrapper_test.sh:
```diff
@@ -3,60 +3,141 @@
+  if [[ "$ret" != "$CORRECT_TMP_FS_MSG" ]]; then
+    got_prefixed=$(prefix_multiline "$ret")
+    fail "Expecting return:\n$expected_tmp_fs_prefixed\nBut got:\n$got_prefixed\n(Calling lkr/fs from lkr)"
+  fi
```
newline

rodrigosiqueira commented on tests/get_maintainer_wrapper_test.sh:
```diff
@@ -3,60 +3,141 @@
+  if [[ "$ret" != "$CORRECT_TMP_MSG" ]]; then
+    got_prefixed=$(prefix_multiline "$ret")
+    fail "Expecting return:\n$expected_tmp_prefixed\nBut got:\n$got_prefixed\n(Calling lkr from lkr/fs)"
+  fi
```
newline

MatheusBernardino commented:
@rodrigosiqueira All changes made! :)
